src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/FallbackProjectManager.cs

```diff
@@ -101,15 +111,35 @@ private void AddFallbackProject(ProjectId projectId, string filePath)
     private void AddFallbackDocument(ProjectKey projectKey, string filePath, string projectFilePath)
     {
         var hostDocument = CreateHostDocument(filePath, projectFilePath);
+        if (hostDocument is null)
+        {
+            return;
+        }
+
         var textLoader = new FileTextLoader(filePath, defaultEncoding: null);
         _projectSnapshotManagerAccessor.Instance.DocumentAdded(projectKey, hostDocument, textLoader);
     }
 
-    private static HostDocument CreateHostDocument(string filePath, string projectFilePath)
+    private static HostDocument? CreateHostDocument(string filePath, string projectFilePath)
     {
-        var targetPath = filePath.StartsWith(projectFilePath, FilePathComparison.Instance)
-            ? filePath[projectFilePath.Length..]
-            : filePath;
+        var projectPath = Path.GetDirectoryName(projectFilePath);
+        if (projectPath is null)
+        {
+            return null;
+        }
+
+        if (!projectPath.EndsWith("\\"))
+        {
+            projectPath += "\\";
+        }
+
+        // The compiler only supports paths that are relative to the project root
+        if (!filePath.StartsWith(projectPath, FilePathComparison.Instance))
```

Comment:
This is the fix. The compiler doesn't support files that aren't in the project cone: https://github.com/dotnet/razor/blob/main/src/Compiler/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectFileSystem.cs#L57


Also the old code had a bug, it was checking StartsWith with the full path to the project file name, which would never have been true.